### PR TITLE
[webui][api] Fix ProjectLogRotate

### PR DESCRIPTION
**Pull request by evanrolfe**
Make a seperate queue for ProjectLogRotate so that the job executes sequentially and never in parallel.

Then we can remove the lock on the events table, since ProjectLogRotate only cares about events.project_logged which is not used by SendEventEmails.

### CODE REVIEW

File: dist/obsapidelayed [hunk 1]
```diff
@@ -64,8 +64,9 @@ case "$1" in
         run_in_api script/delayed_job.api.rb --queue=releasetracking start -i 1000
         run_in_api script/delayed_job.api.rb --queue=issuetracking start -i 1010
         run_in_api script/delayed_job.api.rb --queue=mailers start -i 1020
-        # to be removed later. This is just for processing old jobs without default queue
+        # This is just for processing old jobs without default queue
```
Comment (hennevogel) [inline dist/obsapidelayed]: This is kind of misleading (see issue #3351). Can you change this comment to

```
# The default queue used by ActiveJob (jobs scheduled with .perform_later)
```

File: src/api/app/jobs/project_log_rotate.rb [hunk 2]
```diff
@@ -1,30 +1,21 @@
 class ProjectLogRotate
   def perform
     event_classes = [Event::Package, Event::Project]
+    event_types = event_classes.flat_map(&:descendants).map(&:name)
```
Comment (hennevogel) [inline src/api/app/jobs/project_log_rotate.rb]: Why is this change necessary?

Comment (evanrolfe) [inline src/api/app/jobs/project_log_rotate.rb]: Before it did a `.where` on `Event::Package` and `Event::Project` which does the same thing as this but is harder to work with because you now have two lookup queries as opposed to one. So if I want to fetch 10,000 rows then I can just do `limit(10000)`. The same things get executed but this is just easier to work with.

**Comment by adrianschroeter** [comment]
On Dienstag, 4. Juli 2017, 15:03:46 CEST wrote Henne Vogelsang:
> hennevogel requested changes on this pull request.
> 
> 
> 
> > @@ -64,8 +64,9 @@ case "$1" in
>          run_in_api script/delayed_job.api.rb --queue=releasetracking start -i 1000
>          run_in_api script/delayed_job.api.rb --queue=issuetracking start -i 1010
>          run_in_api script/delayed_job.api.rb --queue=mailers start -i 1020
> -        # to be removed later. This is just for processing old jobs without default queue
> +        # This is just for processing old jobs without default queue
> 
> This is kind of misleading (see issue #3351). Can you change this comment to
> 
> ```
> # The default queue used by ActiveJob (jobs scheduled with .perform_later)

the origin idea behind the comment was that there shouldn't any of these jobs anymore.
Only some after updating from an older instance.

Maybe we should turn this into Admin mails when such a job exist?
It means that it is not assigned to a queue what should not happen anymore.

> ```
> 
> > @@ -1,30 +1,21 @@
>  class ProjectLogRotate
>    def perform
>      event_classes = [Event::Package, Event::Project]
> +    event_types = event_classes.flat_map(&:descendants).map(&:name)
> 
> Why is this change necessary?
> 
> 


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by mdeniz** [comment]
Hey! 😄 
> Make a seperate queue for ProjectLogRotate so that the job executes sequentially and never in parallel.

This is not exactly true, I mean it can be running in parallel. We have one worker for that queue and another without being bounded to any queue (so it can process anything from any queue).

**Comment by hennevogel** [comment]
> the origin idea behind the comment was that there shouldn't any of these jobs anymore.

As noted in #3351 **all** ActiveJob jobs run on this worker, so it might be your original idea but it doesn't apply anymore :-) And it's even worse, we just discovered that they do run there because this worker just grabs **any** job from **any** queue...

**Comment by evanrolfe** [comment]
@hennevogel @mdeniz please review

File: src/api/app/jobs/project_log_rotate.rb [hunk 3]
```diff
@@ -1,30 +1,23 @@
...
-            # Mark the event as logged if the entry was succesfully created
-            processed_ids << event.id if entry.id
-          end
-          Event::Base.where(id: processed_ids).update_all(project_logged: true)
-        end
+    # Create log entries based on the events (but this time, only those in event_classes)
+    Event::Base.where(project_logged: false, eventtype: event_types).find_in_batches(batch_size: 10000) do |events_batch|
+      events_batch.each do |event|
+        entry = ProjectLogEntry.create_from(event)
+        event.update_attributes(project_logged: true) if entry.persisted?
```
Comment (hennevogel) [inline src/api/app/jobs/project_log_rotate.rb]: Why is the if necessary?

File: src/api/app/jobs/send_event_emails.rb [hunk 4]
```diff
@@ -5,18 +5,12 @@ class SendEventEmails
...
   def perform
-    Event::Base.where(mails_sent: false).order(:created_at).limit(1000).lock(true).each do |event|
-      event.mails_sent = true
-      begin
-        event.save!
-      rescue ActiveRecord::StaleObjectError
-        # if someone else saved it too, better don't continue - we're not alone
-        return false
-      end
+    Event::Base.where(mails_sent: false).order(:created_at).limit(1000).each do |event|
```
Comment (hennevogel) [inline src/api/app/jobs/send_event_emails.rb]: I actually think the order should be the other way around. Oldest first.

File: src/api/app/jobs/project_log_rotate.rb (hunk 3 above)
Comment (evanrolfe) [inline src/api/app/jobs/project_log_rotate.rb]: To check that the ProjectLogEntry was saved to the db

File: src/api/app/jobs/send_event_emails.rb (hunk 4 above)
Comment (evanrolfe) [inline src/api/app/jobs/send_event_emails.rb]: The `.order(:created_at)` here is ASC by default so the oldest already come first. But I will add `:asc` to the arguments to make it more clear.

https://apidock.com/rails/ActiveRecord/QueryMethods/order

File: src/api/app/jobs/project_log_rotate.rb (hunk 3 above)
Comment (hennevogel) [inline src/api/app/jobs/project_log_rotate.rb]: Ah, clever

File: src/api/app/jobs/send_event_emails.rb (hunk 4 above)
Comment (hennevogel) [inline src/api/app/jobs/send_event_emails.rb]: Nah, it's okay then

File: dist/obsapidelayed [hunk 5]
```diff
@@ -9,18 +9,18 @@
 #   and its symbolic link
```
Comment (bgeuken) [inline dist/obsapidelayed]: Please update the commit message:-)

File: src/api/app/jobs/send_event_emails.rb [hunk 6]
```diff
@@ -7,7 +7,12 @@ class SendEventEmails
   def perform
     Event::Base.where(mails_sent: false).order(created_at: :asc).limit(1000).each do |event|
       subscribers = event.subscribers
-      next if subscribers.empty?
+
+      if subscribers.empty?
+        event.update_attributes(mails_sent: true)
+        next
+      end
```
Comment (bgeuken) [inline src/api/app/jobs/send_event_emails.rb]: Just a thought, but would it be possible to query all events with mails_sent false and subscribers empty (NULL?) and update them at once?

And then run the mails_sent false but subscribers not empty in a separate query. I don't have any numbers to back this up, but I could imagine this (the first query) would be faster.

Comment (bgeuken) [inline src/api/app/jobs/send_event_emails.rb]: Of course this would make it harder to limit it to 1000, if this is critical ...

**Comment by bgeuken** [comment]
LGTM

File: src/api/app/jobs/send_event_emails.rb (hunk 6 above)
Comment (hennevogel) [inline src/api/app/jobs/send_event_emails.rb]: subscribers is unfortunately not an association...

File: src/api/app/jobs/send_event_emails.rb [hunk 7]
```diff
@@ -5,18 +5,17 @@ class SendEventEmails
...
-      next if subscribers.empty?
-      EventMailer.event(subscribers, event).deliver_now
+
+      if subscribers.empty?
+        event.update_attributes(mails_sent: true)
+        next
+      end
+
       create_rss_notifications(event)
+      EventMailer.event(subscribers, event).deliver_now
```
Comment (mdeniz) [inline src/api/app/jobs/send_event_emails.rb]: What we already talked... but at least to have it here.
We should have lines 16 and 17 in a begin-rescue block to make sure that a failing event is marked as sent (like it was before). To not get stuck on events unprocessed over and over again

File: src/api/app/jobs/send_event_emails.rb (hunk 6 above)
Comment (evanrolfe) [inline src/api/app/jobs/send_event_emails.rb]: Yes unfortunately the logic for determine the subscribers for an event is very complex and determined by this class and the logic only works one way: we can find subscribers for an event, but not events for a subscriber (or no subscriber).

https://github.com/evanrolfe/open-build-service/blob/fix%2Fproject_log_rotate/src/api/app/models/event_find_subscriptions.rb#L1

Comment (hennevogel) [inline src/api/app/jobs/send_event_emails.rb]: @evanrolfe can you please create a backlog entry for changing this to an association?

**Comment by hennevogel** [comment]
Okay let's do this!

giphy

File: src/api/app/jobs/send_event_emails.rb (hunk 6 above)
Comment (evanrolfe) [inline src/api/app/jobs/send_event_emails.rb]: @hennevogel what would the association be? A many-to-many between events and subscriptions?

**Comment by hennevogel** [comment]
[Fix DelayedJob/Clockword setup](https://trello.com/c/U6apQGgR/450-fix-delayedjob-clockword-setup)

File: src/api/app/jobs/send_event_emails.rb (hunk 6 above)
Comment (hennevogel) [inline src/api/app/jobs/send_event_emails.rb]: @evanrolfe I don't know but not this 💩

Comment (evanrolfe) [inline src/api/app/jobs/send_event_emails.rb]: @hennevogel https://trello.com/c/rgzkWKbN/380-refactor-eventfindsubscriptions
